### ISSUE TITLE
Test Mac build with embedded zcashd

### ISSUE THREAD

**Issue by adityapk00** [ZcashFoundation/zec-qt-wallet#49]
Test mac dmg package. 

(Putting dmg in a zip otherwise github won't let me upload it to an issue)
[MacOS-zec-qt-wallet-v0.3.1.dmg.zip](https://github.com/adityapk00/zec-qt-wallet/files/2581915/MacOS-zec-qt-wallet-v0.3.1.dmg.zip)

**Comment by garethtdavies** [comment]
Closing `zec-qt-wallet` does not close `zcashd`. This is repeatable, clicking on the close icon just closes the application and doesn't give me the message as shown on Linux. Confirmed that zcashd is indeed still running as a daemon.

**Comment by adityapk00** [comment]
> Closing `zec-qt-wallet` does not close `zcashd`. This is repeatable, clicking on the close icon just closes the application and doesn't give me the message as shown on Linux. Confirmed that zcashd is indeed still running as a daemon.

Hmm... I can't seem to repro this. Are you closing by clicking the red 'x' or cmd-Q? Also, what version of MacOS are you on?

**Comment by rex4539** [comment]
Actually, I am able to test a few things while zcashd is indexing :)

The DMG reads `MacOS-zec-qt-wallet-v0.3.1` but the mounted image is named `zec-qt-wallet-v0.3.0`

**Comment by garethtdavies** [comment]
Ah, I know what's going on. On a rebooted machine this works. If I crash the app (by double clicking address book as above - as I don't have the fixed version) then restarting zec-qt-wallet it thinks it is not running the embedded zcashd so doesn't close it. I like @rex4539's idea of asking to close the daemon or keep it running 👍

**Comment by adityapk00** [comment]
Ah, I get it. I wonder if I can detect a crashed zec-qt-wallet and then reconnect to zcashd. I'll look into it.

**Comment by garethtdavies** [comment]
@rex4539 in the Monero example how do you close the daemon in the case you leave it running? As in, is there an easy way to do it other than shutting down the machine / killing the process manually? Can you re-open the GUI and then if you close it'll ask you again or does it behave like the current zec-qt-wallet and won't do anything if it detects an already running instance?

**Comment by rex4539** [comment]
Yes, you can relaunch the GUI, close it again and this time choose to shut down the daemon.

**Comment by rex4539** [comment]
After reindexing finished, I'm trying to start the zec-qt-wallet from the DMG and I get this error.

![screen shot 2018-11-15 at 21 37 02](https://user-images.githubusercontent.com/227442/48577346-a79ad400-e91f-11e8-834c-4359650061d9.png)

The one I compile from source starts fine.

**Comment by adityapk00** [comment]
Hmm. I can't repro this either. Maybe it has something to do with the txindex.

I'm building a node with txindex to test it out.

**Comment by garethtdavies** [comment]
I got that first time too but I put it down to my existing setup. I think I moved my old zcash.conf file and did actually delete the params directory (so I could see the loading process) and it all worked fine.

Presumably, anything that prevents a clean zcashd startup such as anything that requires a reindex would cause this?

Can you display the error message from zcashd as that would likely be much more informative?

**Comment by rex4539** [comment]
debug.log doesn't log a single error.

zcashd actually continues to run just fine.

**Comment by rex4539** [comment]
So for some reason zec-qt-wallet thinks zcashd could not be started although it runs fine in the background.

If I quit zec-qt-wallet and relaunch it then it connects to the same zcashd fine.

**Comment by adityapk00** [comment]
Here's a new build of the dmg. It has all of yesterday's bug fixes and some additional debugging info. 

By the way, I really appreciate your help testing this and the very detailed feedback you are providing. It is super valuable!

[MacOS-zec-qt-wallet-v0.3.1.dmg.zip](https://github.com/adityapk00/zec-qt-wallet/files/2586881/MacOS-zec-qt-wallet-v0.3.1.dmg.zip)

**Comment by rex4539** [comment]
Nothing again in debug.log

This time I get an extra error window.

![screen shot 2018-11-15 at 22 18 57](https://user-images.githubusercontent.com/227442/48579370-f434de00-e924-11e8-99d5-0cff675d627e.png)

After clicking "ok" I get the previous error window too.

**Comment by adityapk00** [comment]
I think I figured out what's going on with the crash. If the params change between zcash invokations, zcashd exits unexpectedly. 

For me:
- start zcashd manually with -txindex=1 as a command line option (not in zcash.conf)
- stop zcashd
- start zec-qt-wallet, allowing embedded zcashd to start.

Now what happens is since -txindex=1 is not in zcash.conf, zcashd tries to start without it, and errors out with `You need to rebuild the database using -reindex to change -txindex` , resulting in an unexpected exit from zcashd. 

Is this happening to you as well? Did you pass -txindex=1 on the command line (and not in zcash.conf)?

**Comment by adityapk00** [comment]
[MacOS-zec-qt-wallet-v0.3.2.dmg.zip](https://github.com/ZcashFoundation/zec-qt-wallet/files/2587707/MacOS-zec-qt-wallet-v0.3.2.dmg.zip)

Really interesting issue. I added an error msg that hopefully gives users more context.

**Comment by garethtdavies** [comment]
That sounds like a great approach.

How are you checking the parameters (as in zcash params) existence and correctness? I did delete mine which fixed the message but did a few other changes too so hard to know if that is related to the issue. Any reason why suggesting these are corrupt is the message shown above any other startup error?

**Comment by adityapk00** [comment]
The most common reason this happened for me was that the download of the params failed for some reason and zcashd exits. Right now, zec-qt-wallet only checks that the params exist, and if they do, it moves on. Maybe I should do a sha256 check on the params files. But that will slow down startup quite a bit. 

With this new approach, whatever is wrong, we can rely on zcashd reporting it to the user, so they can get more context.

**Comment by garethtdavies** [comment]
That makes sense. I thought there might be something else going on which was causing an error with existing params not downloaded via zec-qt-wallet but that's clearly not the case. If they were missing/corrupt zcashd would report that anyway on startup so that case is covered.

**Comment by rex4539** [comment]
> I think I figured out what's going on with the crash. If the params change between zcash invokations, zcashd exits unexpectedly.
> 
> For me:
> 
> * start zcashd manually with -txindex=1 as a command line option (not in zcash.conf)
> * stop zcashd
> * start zec-qt-wallet, allowing embedded zcashd to start.
> 
> Now what happens is since -txindex=1 is not in zcash.conf, zcashd tries to start without it, and errors out with `You need to rebuild the database using -reindex to change -txindex` , resulting in an unexpected exit from zcashd.
> 
> Is this happening to you as well? Did you pass -txindex=1 on the command line (and not in zcash.conf)?

No, the exact opposite is occurring. I do not pass `-txindex=1` on the command line but I have it in `zcash.conf`

**Comment by adityapk00** [comment]
Interesting. Do you see any error messages from zcashd in the new build?

**Comment by rex4539** [comment]
No error messages.

**Comment by rex4539** [comment]
Just to make it clear, the embedded zcashd does not crash and runs just fine during the whole process.

**Comment by adityapk00** [comment]
OK, I'm officially baffled :) 

Can you paste your zcash.conf? What version of MacOS are you on? zcashd version?

I'm going to try to duplicate the entire setup. 

As a backup plan, I think I can see if there's a zcashd running even if the process returns an error, (i.e., detect this exact situation) and just not show the error.

**Comment by garethtdavies** [comment]
I just tried on a new Mac. First error screen showed the expected zcashd warning that the params were missing but then zec-qt-wallet showed a couldn't start the embedded zcashd error with The process returned: "Unknown error"

**Comment by rex4539** [comment]
I run zcashd via a local tor node.

```
addnode=mainnet.z.cash
addnode=zcmaintvsivr7pcn.onion
addnode=zcashiqykswlzpsu.onion
addnode=zcashqhrmju6zfhn.onion
addnode=zcashgmvxwrmjsut.onion
addnode=zcashz3uma65ix7b.onion
addnode=zcashiyf4kxluf3x.onion
addnode=zcashwfe4x3jkz2b.onion
addnode=zcashvkeki52iqpc.onion
addnode=zcasha3cmfrpy7b7.onion
addnode=zcashz7ed3nvbdxm.onion
addnode=zcash5adwfpxfuvf.onion
addnode=zcashixg5ol2ndo4.onion
addnode=zcashuzwa365oh3n.onion
addnode=zcashskbeoiwtym3.onion
addnode=zcashuyvk5e7qfzy.onion
addnode=fhsxfrwpyrtoxeal.onion
addnode=zcash2iihed2wdux.onion
addnode=w3dxku36wbp3lowx.onion
addnode=zcashuhmzycmlwld.onion
rpcuser=username
rpcpassword=XXXXXXXXXXX
daemon=1
showmetrics=0
gen=0
proxy=127.0.0.1:9050
onlynet=onion
experimentalfeatures=1
paymentdisclosure=1
debug=paymentdisclosure
txindex=1
```

**Comment by garethtdavies** [comment]
I then renamed an existing zcash directory at `~/Library/Application Support/zcash` and did nothing else and this time `zec-qt-wallet` started up properly and downloaded the parameters so it's the existing stuff that's causing the issues.

**Comment by adityapk00** [comment]
So zec-qt-wallet didn’t download the params? Maybe that’s what causing this.

**Comment by garethtdavies** [comment]
For me, yes. As now when I rename the old folder back to Zcash it works so initial issue definitely related to the params.

**Comment by garethtdavies** [comment]
And finally I tried on a third Mac that has never seen Zcash before. It worked as intended so something is causing issues with existing installs I imagine.

**Comment by rex4539** [comment]
The funny this is that the error appears only when the embedded zcashd is started.

After you get the error, if you quit zec-qt-wallet and restart it, it connects to the embedded zcashd just fine.

So something is going on when the embedded zcashd is started.

**Comment by adityapk00** [comment]
One more attempt with the latest fix. Unfortunately, I still can't repro this on my Mac Mini, so go to plan B. This time, we double check to see if zcashd has escaped into the background somehow, and try to connect to it before we give up and show an error. 
[MacOS-zec-qt-wallet-v0.3.2a.dmg.zip](https://github.com/ZcashFoundation/zec-qt-wallet/files/2592300/MacOS-zec-qt-wallet-v0.3.2a.dmg.zip)

**Comment by rex4539** [comment]
No change here. Same error message appears immediately upon launch.

**Comment by adityapk00** [comment]
Well, that's disappointing. And if you restart zec-qt-wallet, it will connect to the embedded zcashd that it started last time and continue?

**Comment by rex4539** [comment]
Yes.

**Comment by adityapk00** [comment]
[MacOS-zec-qt-wallet-v0.3.2b.dmg.zip](https://github.com/ZcashFoundation/zec-qt-wallet/files/2592311/MacOS-zec-qt-wallet-v0.3.2b.dmg.zip)
Ok, one more, last, desperate attempt.

**Comment by rex4539** [comment]
Unfortunately, still getting the same error.

**Comment by adityapk00** [comment]
Ok, I'm gonna go bang my head against the wall to see if that helps.

**Comment by rex4539** [comment]
Don’t despair! We’ll get to the bottom of it :)

**Comment by adityapk00** [comment]
Ok. So I’ve spent half a day on this again, but can’t replicate it. I’m using your exact zcash.conf and still nothing. 

So here’s the plan. I’ll update the error message to something like: “try restarting, if that doesn’t work, run external zcashd”. 

We’ll go ahead with the macOS release tomorrow and see if more people run into it and if we can get more info. 

Thanks a ton to both of you for the extensive testing and helping get macOS ready.

**Comment by garethtdavies** [comment]
Also see #58

**Comment by rex4539** [comment]
https://www.reddit.com/r/zec/comments/9yqlfl/comment/ea4pgrx?st=JOR5GOI5&sh=da6af729

**Comment by rex4539** [comment]
ok, I had some free time today and figured this out.

The problem appears when zcash.conf contains the entry `daemon=1`.

If this entry is not there, everything works perfectly.

According to the [documentation](https://zcash.readthedocs.io/en/latest/rtd_pages/zcash_mining_guide.html#mining), this allows to "run it in the background (without the node metrics screen that is normally displayed)".

My understating is that the embedded zcashd forces the metrics screen to be loaded so if the `daemon=1` flag is passed, it conflicts with the zec-qt-wallet code and the error is thrown.

In any case, we now know what causes the bug. Let's figure out the best way to fix it.

Obvious solution would be to inform the user to remove this entry from zcash.conf

Another solution would be to automatically remove the entry from zcash.conf as soon as zec-qt-wallet is launched.

I have no idea how this entry ended up in my zcash.conf file but my guess would be that I have either entered it manually at some point when I was testing something (more unlikely since it happens to other users on Reddit too) or zcash4mac (or some other GUI) forced this entry upon first installation. So if the user ever ran this software in the past, have the `daemon=1` entry in their zcash.conf

**Comment by rex4539** [comment]
Of course, the best fix would be to make it work with the daemon spawned :)

**Comment by adityapk00** [comment]
Ooooohhh. I get it now. Yes, you are right, the `daemon=1` will push zcashd into the background, and zec-qt-wallet will get confused. 

This is amazing. Thanks so much, @rex4539 for your impressive detective work. I think I can fix this quite easily now that I understand what's happening.

**Comment by rex4539** [comment]
Verified as fixed.

**Comment by adityapk00** [comment]
Fixed in releae v0.4.2
